### PR TITLE
Render table in the shadow DOM for `@finos/perspective-viewer-datagrid`

### PR DESCRIPTION
**Pull request by tomjakubowski**
Move the `<regular-table>` element of datagrid into the shadow DOM by default.

This change isolates the regular-table from global styles, so CSS rules on `regular-table` or `table` no longer affect datagrid. Style rules on the `regular-table` can be ported using the part named `regular-table`

```
// old style
perspective-viewer-datagrid regular-table {
  font-family: "Comic Sans MS";
}
// shadow dom
perspective-viewer-datagrid::part(regular-table) {
  font-family: "Comic Sans MS";
}
```

To disable this feature and continue to use light DOM rendering, set:

```
HTMLPerspectiveViewerDatagridPluginElement.renderTarget = "light"
```

### CODE REVIEW

**Review by texodus — CHANGES_REQUESTED** [review]
Thanks for the PR! A couple of things need fixing:

1. I'm ok with the configuration flag being a global static on the `perspective-viewer-datagrid` element, because there is no global settings feature yet, but one is obviously needed. However, let's please rename this to `enableShadowDOM`.
2. `enableShadowDOM` should be `true` by default, e.g. we should default use the Shadow DOM. It looks like the examples work in this mode with this PR (👍 ), so no changes needed there. It is ok if the tests remain as Light DOM tests because they do some DOM surgery to check table elements, but ...
3. There needs to be at least a few tests added that test Shadow DOM mode specifically. If it's doable, I'd really like to see _all_ the tests run in parameterized in both modes, but this may be a lot of work (?)
4. The tests are currently failing regardless (looks like `perspective-jupyterlab`).
5. Let's remove the `examples/blocks` code which parses the URL.

File: packages/perspective-viewer-datagrid/src/js/custom_elements/datagrid.js [hunk 1]
```diff
@@ -12,26 +12,61 @@
...
+        this.#parentObserver = new MutationObserver((mutations) => {
+            for (let mut of mutations) {
+                if (
+                    mut.type === "attributes" &&
+                    mut.attributeName === "settings"
+                ) {
+                    this.#syncAttributes();
+                }
+            }
+        });
```
[the remark this answers is not in the document]
Comment (tomjakubowski) [reply packages/perspective-viewer-datagrid/src/js/custom_elements/datagrid.js:42]: Yeah, it's also what's breaking the snapshot tests (only on Linux, though!). I think :host-context should work, I will try that

File: packages/perspective-viewer-datagrid/src/js/custom_elements/datagrid.js [hunk 2]
```diff
@@ -12,26 +12,61 @@
...
 import { save } from "../plugin/save";
 import { draw } from "../plugin/draw";
 import getDefaultConfig from "../default_config.js";
+import datagridStyles from "../../../dist/css/perspective-viewer-datagrid.css";
 
 /**
  * The custom element class for this plugin.  The interface methods for this
  */
-export class HTMLPerspectiveViewerDatagridPluginElement extends HTMLElement {
+export const HTMLPerspectiveViewerDatagridPluginElement = class extends HTMLElement {
```
[the remark this answers is not in the document]
Comment (tomjakubowski) [reply packages/perspective-viewer-datagrid/src/js/custom_elements/datagrid.js:23]: not a clue where that came from, weird! Maybe an accidental auto-fix in VSCode or something

File: examples/blocks/src/magic/index.js [hunk 3]
```diff
@@ -117,4 +118,10 @@ class MagicApp extends HTMLElement {
     }
 }
 
+async () => {
+    await window.customElements.whenDefined("perspective-viewer-datagrid");
+    window.customElements.get(
+        "perspective-viewer-datagrid"
+    ).experimentalEnableShadowDom = true;
+};
```
[the remark this answers is not in the document]
Comment (tomjakubowski) [reply examples/blocks/src/magic/index.js:126]: yeah leftover from debugging

**Comment by tomjakubowski** [comment]
> I'm ok with the configuration flag being a global static on the perspective-viewer-datagrid element, because there is no global settings feature yet, but one is obviously needed. However, let's please rename this to enableShadowDOM.

Another API I was thinking about was sticking it as an attribute/flag on `<perspective-viewer>`, like `<perspective-viewer shadow />`. A problem is I imagine it would be disastrous to update a particular element to use/not use shadow DOM at runtime, i.e. to toggle usage of shadow DOM in response to changes to that attribute.

> There needs to be at least a few tests added that test Shadow DOM mode specifically. If it's doable, I'd really like to see all the tests run in parameterized in both modes, but this may be a lot of work (?)

I'll try this, it's a good idea to rerun the same tests with shadow DOM enabled/disabled. And I'll think about a test or two just for shadow DOM

**Comment by tomjakubowski** [comment]
All the checks are passing now, re-review?

File: packages/perspective-viewer-datagrid/src/js/custom_elements/datagrid.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (tomjakubowski) [reply packages/perspective-viewer-datagrid/src/js/custom_elements/datagrid.js]: updated

File: tools/perspective-test/results.tar.gz
Comment (tomjakubowski) [inline tools/perspective-test/results.tar.gz:1]: curious if you have any recommendations for quickly diffing these tarballs?

Comment (texodus) [reply tools/perspective-test/results.tar.gz:1]: Not sure what you mean. In theory, if the snapshots don't change, the tarball should be identical. We've added a few [flags to the gzip process](https://github.com/finos/perspective/blob/master/tools/perspective-test/src/js/global_teardown.ts#L35) to force this to be the case regardless of things like file creation time.

I have noticed that this is not always the case, but I don't know why yet, nor have I even validated that this is an issue with the _tarball_ or the actual files.

**Review by texodus — APPROVED** [review]
Thanks for the PR! Looks good!

**Comment by jakehadar** [comment]
> > I'm ok with the configuration flag being a global static on the perspective-viewer-datagrid element, because there is no global settings feature yet, but one is obviously needed. However, let's please rename this to enableShadowDOM.
> 
> Another API I was thinking about was sticking it as an attribute/flag on `<perspective-viewer>`, like `<perspective-viewer shadow />`. A problem is I imagine it would be disastrous to update a particular element to use/not use shadow DOM at runtime, i.e. to toggle usage of shadow DOM in response to changes to that attribute.
> 
> > There needs to be at least a few tests added that test Shadow DOM mode specifically. If it's doable, I'd really like to see all the tests run in parameterized in both modes, but this may be a lot of work (?)
> 
> I'll try this, it's a good idea to rerun the same tests with shadow DOM enabled/disabled. And I'll think about a test or two just for shadow DOM

Hi, was this attribute flag ever exposed publicly? Looks like it is a static/private attribute set at runtime based on browser capaibility  https://github.com/finos/perspective/blame/84605851230c3dd1cc58652b47babd04cc9fcc31/packages/perspective-viewer-datagrid/src/js/custom_elements/datagrid.js#L141 
This introduces inconsistencies between browsers and i believe the shadow rendering policy should be something the consumer can opt-out of

**Comment by tomjakubowski** [comment]
@jakehadar The inferred setting is a default. You can override it by using [`customElements.whenDefined()`](https://developer.mozilla.org/en-US/docs/Web/API/CustomElementRegistry/whenDefined) to access the custom element constructor when it is added to the registry, and then set the `renderTarget` property on it to whichever you prefer.

```
customElements.whenDefined("perspective-viewer-datagrid").then((datagrid) => {
  console.log("setting datagrid render target");
  datagrid.renderTarget = "light";
});
```

`datagrid.renderTarget` is currently an undocumented API, but I don't anticipate it will change in the near term.

**Comment by jakehadar** [comment]
Awesome, this is exactly what I need. Thank you for the quick response!Sent from my iPhoneOn May 8, 2024, at 14:01, Tom Jakubowski ***@***.***> wrote:﻿
@jakehadar The inferred setting is a default. You can override it by using customElements.whenDefined() to access the custom element constructor when it is added to the registry, and then set the renderTarget property on it to whichever you prefer.
customElements.whenDefined("perspective-viewer-datagrid").then((datagrid) => {
  console.log("setting datagrid render target");
  datagrid.renderTarget = "light";
});

datagrid.renderTarget is currently an undocumented API, but I don't anticipate it will change in the near term.

**Pull request merged**